libp2p/autotls/acme/api.nim

```diff
@@ -0,0 +1,342 @@
+import options, base64, sequtils, serialization, json_serialization, tables
+from times import DateTime, parse
+import chronos/apps/http/httpclient, jwt, results, bearssl/pem
+from std/json import
+  JsonNode, `%*`, `%`, `[]`, `[]=`, `$`, parseJson, getStr, items, newJObject, to,
+  JsonKindError
+
+import ../utils
+import ../../crypto/crypto
+import ../../crypto/rsa
+import ../../transports/tls/certificate_ffi
+import ../../transports/tls/certificate
+
+const
+  LetsEncryptURL* = "https://acme-v02.api.letsencrypt.org"
+  LetsEncryptURLStaging* = "https://acme-staging-v02.api.letsencrypt.org"
+  Alg = "RS256"
+  DefaultChalCompletedRetries = 10
+  DefaultChalCompletedRetryTime = 1.seconds
+  DefaultFinalizeRetries = 10
+  DefaultFinalizeRetryTime = 1.seconds
+  DefaultRandStringSize = 256
+
+type Nonce = string
+type Kid = string
+type AccountStatus = string
+type ACMERequestHeader = JsonNode
+
+type ACMEDirectory = object
+  newNonce: string
+  newOrder: string
+  newAccount: string
+
+type ACMEApi* = object
+  directory: ACMEDirectory
+  session*: HttpSessionRef
+  acmeServerURL: string
+
+type ACMEResponse = object
+  body: JsonNode
+  headers: HttpTable
+
+type ACMERegisterResponse* = object
+  kid*: Kid
+  status*: AccountStatus
+
+type ACMEChallenge = object
+  url: string
+  `type`: string
+  status: string
+  token: string
+
+type ACMEAuthorizationsResponse* = object
+  challenges: seq[ACMEChallenge]
+
+type ACMEChallengeResponse* = object
+  finalizeURL: string
+  orderURL: string
+  authzURL: string
+  dns01: ACMEChallenge
+
+type ACMECompletedResponse* = object
+  checkURL: string
+
+type ACMECheckResponse* = object
+  status: string
+
+type ACMEFinalizedResponse* = object
+
+type ACMEOrderResponse* = object
+  certificate: string
+  expires: string
+
+type ACMECertificateResponse* = object
+  rawCertificate: string
+  certificateExpiry: DateTime
+
+template wrapSerialization(body: untyped): untyped =
+  try:
+    body
+  except JsonKindError as exc:
+    raise newException(ACMEError, "Failed to decode JSON", exc)
+  except exceptions.KeyError as exc:
+    raise newException(ACMEError, "Failed to decode JSON", exc)
+  except ValueError as exc:
+    raise newException(ACMEError, "Failed to decode JSON", exc)
+  except HttpError as exc:
+    raise newException(ACMEError, "Failed to connect to ACME server", exc)
```

Comment:
interesting!

- name is not ideal, perhaps "handle error"?
- when errors are raised using this, it will be impossible to tell which place it raised as there are may places with same error message
- why is it used on some places and on other it isn't? can we have have unified usage?